Refactor workflow engine to be able to serialize snakefiles on the fly

DanInci opened the pull request:
Refactor the workflow engine code, such that the `Snakemake` file is generated on the fly.
The following generic interface is provided for workflow execution:

```
class WorkflowEngine:
    """Interface for the workflow engine."""

    def run_workflow(self, benchmark: Benchmark, cores: int = 1, dryrun: bool = False, work_dir: str = os.getcwd(), **kwargs):
        """
        Serializes & runs benchmark workflow.

        Args:
            benchmark (Benchmark): benchmark to run
            cores (int): number of cores to run
            dryrun (bool): validate the workflow with the benchmark without actual execution
            work_dir (str): working directory
            **kwargs: keyword arguments to pass to the workflow engine

        Returns:
        - Status code of the workflow run.
        """
        raise NotImplementedError("Method not implemented yet")

    def serialize_workflow(self, benchmark: Benchmark, output_dir: str = os.getcwd()):
        """
        Serializes a workflow file for the benchmark.

        Args:
            benchmark (Benchmark): benchmark to serialize
            output_dir (str): output directory for the workflow file

        Returns:
        - Workflow file path.
        """
        raise NotImplementedError("Method not implemented yet")

    def run_node_workflow(self, node: BenchmarkNode, cores: int = 1, dryrun: bool = False, work_dir: str = os.getcwd(), **kwargs):
        """
        Serializes & runs benchmark node workflow.

        Args:
            node (Benchmark): benchmark node to run
            cores (int): number of cores to run
            dryrun (bool): validate the workflow with the benchmark without actual execution
            work_dir (str): working directory
            **kwargs: keyword arguments to pass to the workflow engine

        Returns:
        - Status code of the workflow run.
        """
        raise NotImplementedError("Method not implemented yet")

    def serialize_node_workflow(self, node: BenchmarkNode, output_dir: str = os.getcwd()):
        """
        Serializes a workflow file for a benchmark node.

        Args:
            node (BenchmarkNode): benchmark node to serialize
            output_dir (str): output directory for the workflow file

        Returns:
        - Workflow file path.
        """
        raise NotImplementedError("Method not implemented yet")
```

DanInci committed:
Refactor workflow engine. Serialize Snakefile on the fly.

DanInci committed:
Serialize a node workflow to a Snakefile

DanInci committed:
Write documentation for the workflow engine interface.

DanInci committed:
Update WorkflowEngine interface

DanInci committed:
Some fixes

almutlue commented:
Two quick comments (proper code review will follow): 

1. Type hints and docstrings would really help code readability
2. There are some controversies around using pickle (mainly for [security](https://realpython.com/python-pickle-module/#security-concerns-with-the-python-pickle-module) reasons, but also because objects will not be compatible across versions/systems). I wonder about the advantage of using it here? Why not starting with the `benchmark.yaml` as input and initiate the benchmark/node class from there?

DanInci commented:
> Two quick comments (proper code review will follow):
> 
> 1. Type hints and docstrings would really help code readability
> 2. There are some controversies around using pickle (mainly for [security](https://realpython.com/python-pickle-module/#security-concerns-with-the-python-pickle-module) reasons, but also because objects will not be compatible across versions/systems). I wonder about the advantage of using it here? Why not starting with the `benchmark.yaml` as input and initiate the benchmark/node class from there?

1. I agree, will ad them.
2. Was not aware about the security issues. Thanks for sharing. My plan was to design the interface such that it accepts the already constructed and validated benchmark model and uses it in the workflow engine. The pickeling was to avoid deconstructing & reconstructing the model again, but there's no other harm in it, other than doing that twice.

DanInci committed:
Instead of pickling model, try recreating it benchmark model.

DanInci committed:
Some fixes

DanInci committed:
Implement running of the snakemake workflows

DanInci committed:
Add some more inferred types and documentation to workflow engine

almutlue commented on src/workflow/snakemake/format/formatter.py:
```diff
@@ -12,16 +17,43 @@ def format_output_templates_to_be_expanded(node):
 
 
-def extract_stages_from_path(path, known_stages):
+def format_input_templates_to_be_expanded(benchmark: Benchmark, wildcards):
```
__Question__: What is the type of wildcards? Is there a class that defines which attributes (`pre`, `post`, `name`) belong to it?

DanInci replied:
These [wildcards](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#wildcards) come from the snakemake world and they are an implementation details.

Based on the current implementation, they should have this type. `pre`, `post`, `name` are syntax wildcards that should exist in the input format for different rules, based on which snakemake automatically knows what rules to link.
```
class Wildcards(NamedTuple):
    pre: str
    post: str
    name: str
```

almutlue commented on src/workflow/snakemake/format/formatter.py:
```diff
@@ -12,16 +17,43 @@ def format_output_templates_to_be_expanded(node):
+    return inputs
+
+
+def _extract_stages_from_path(path: str, known_stage_ids: Set[str]):
```
__Question__: In general this looks like a brittle logic. Maybe I misunderstand the purpose, but why do you use `while` loops ans indexes, instead of directly matching the sub parts of the path with the known_stage_id?
Something like:
`stages = [part for part in parts if part in known_stage_ids]`

I think the stage dependencies should also be encoded in the dag. It would probably be safer to extract them from there?

DanInci replied:
I agree the logic is brittle. Unfortunately I left it as is after I experimented a bit with the algorithm.
The idea was to collect tuples of (stage_id, module_id, param_id) from the prefix path of the inputs, not just the stage names.
i.e. `[('data', 'D1', 'default'), ('methods', 'M1', 'default')]`

But I can always retry and write this in a more functional and simpler way.

almutlue commented on src/workflow/snakemake/format/formatter.py:
```diff
@@ -12,16 +17,43 @@ def format_output_templates_to_be_expanded(node):
+
+
+def _extract_stages_from_path(path: str, known_stage_ids: Set[str]):
     parts = path.split('/')
```
```suggestion
    parts = [p for p in Path(path).parts]
```
In general, it is a good practice to use `from pathlib import Path` for path operations as it is system independent.

almutlue commented on src/workflow/snakemake/format/formatter.py:
```diff
@@ -12,16 +17,43 @@ def format_output_templates_to_be_expanded(node):
+
+    node_hash = hash(BenchmarkNode.to_id(stage_id, module_id, param_id, after_stage))
+    matching_node = next((node for node in nodes if hash(node) == node_hash), None)
+    assert matching_node is not None
```
__Fix__: It is better to raise an error with a clear error message, than using `assert`. I think that is mainly used for tests

DanInci replied:
Agree, took some shortcuts during initial development, and originally the asserts were just to make sure i'm covering all the corner cases. At this point they can either be entirely removed or transformed into an error. I will make them a runtime error.

almutlue commented on src/workflow/snakemake/format/formatter.py:
```diff
@@ -12,16 +17,43 @@ def format_output_templates_to_be_expanded(node):
+    name = wildcards.name
+
+    nodes = benchmark.get_nodes()
+    stage_ids = set([node.stage_id for node in nodes])
```
__Question__: Can there be duplicates returned by `benchmark.get_nodes()`? It is probably better to deduplicate at the class methods level instead of always using `set` downstream.

DanInci replied:
Yes, the nodes here represent benchmarking nodes, specific for a `stage_id / module_id / params` combination. As you can image there might be multiple nodes for each stage id. And the `get_nodes()` method has multiple purposes. Here I was trying to get the unique stage_ids in the benchmark, but you have a point, and I could just transform this into a separate method in the `Benchmark` class.

almutlue commented on src/workflow/snakemake/format/formatter.py:
```diff
@@ -35,7 +67,7 @@ def extract_stages_from_path(path, known_stages):
 
 
-def match_node_format(to_match):
+def _match_node_format(to_match: str):
```
__Fix__: If `to_match` can be something else than a string this should be reflected in the type e.g. `to_match: Union[str, tuple]`

almutlue commented on src/workflow/snakemake/format/formatter.py:
```diff
@@ -35,7 +67,7 @@ def extract_stages_from_path(path, known_stages):
+def _match_node_format(to_match: str):
     if type(to_match) is str:
         to_match = to_match.split('/')
 
```
__Fix__: Similar to above, this logic seems brittle as it makes a lot of positional assumptions. Why extracting those information from the string and not from one of the classes, e.g. the node class? This makes it also much safer with regards to structural changes upstream.

almutlue commented on src/workflow/snakemake/format/formatter.py:
```diff
@@ -68,7 +100,7 @@ def match_input_module(input, stages, name):
         return None
```
```suggestion
```
You usually don't need to explicitly return `None`

almutlue commented on src/converter/linkml_converter.py:
```diff
@@ -1,15 +1,23 @@
+    def __init__(self, benchmark_file):
         super().__init__()
-        self.benchmark = benchmark
+        self.benchmark_file = os.path.abspath(benchmark_file)
```
__Fix(code duplication)__: If you assign something as attribute, there is no reason to also provide a method that calls the attribute. Maybe this comes from the `Converter` class, but why not defining it as attribute there already.

DanInci replied:
Fair point, the main reason was overloading, and I was just following the pattern, even though the attribute was already declared in the child class. Regardless, I think most of this code will be simplified once we remove the `YamlConverter` and get rid of the unnecessary hierarchy. I could do this already, but I don't want to overlap with your refactoring?

almutlue commented on src/converter/linkml_converter.py:
```diff
@@ -1,15 +1,23 @@
         super().__init__()
-        self.benchmark = benchmark
+        self.benchmark_file = os.path.abspath(benchmark_file)
+        self.benchmark = load_yaml(benchmark_file)
```
See comment above

almutlue reviewed (commented):
Thanks a lot for this very extensive PR. 

I added some questions/suggestions. They are mainly about reducing duplication to keep the code simple and clean. As more general comments:

1. Why not starting from the `benchmark.yaml` file and initialize the `WorkflowEnginge` from there? This way we could also skip reassigning the `benchmark.yaml` file as `Benchmark/Converter` attribute. 
2. Raise Errors with a clear error message instead of using `assert`

3. Type hints are also used to describe return values, e.g. `def func(arg:argtype) -> returntype:`. I think code readability would be much easier.
4. There are quite some places were the code looks brittle and will probably fail upon small changes, e.g. in the data schema or the path description. I tried to point them out, but probably missed some. In general it is good to e.g. use named attributes instead of positional calls. I would also try to avoid string manipulation as much as possible.

DanInci commented:
> Thanks a lot for this very extensive PR.
> 
> I added some questions/suggestions. They are mainly about reducing duplication to keep the code simple and clean. As more general comments:
> 
> 1. Why not starting from the `benchmark.yaml` file and initialize the `WorkflowEnginge` from there? This way we could also skip reassigning the `benchmark.yaml` file as `Benchmark/Converter` attribute.
> 2. Raise Errors with a clear error message instead of using `assert`
> 3. Type hints are also used to describe return values, e.g. `def func(arg:argtype) -> returntype:`. I think code readability would be much easier.
> 4. There are quite some places were the code looks brittle and will probably fail upon small changes, e.g. in the data schema or the path description. I tried to point them out, but probably missed some. In general it is good to e.g. use named attributes instead of positional calls. I would also try to avoid string manipulation as much as possible.

1. My intuition tells me that the workflow engine should be called after the benchmark file has been validated, and the benchmark model constructed. This is reflected by the interface to the module that accepts only `benchmark: Benchmark` or `node: BenchmarkNode`. The alternative allows direct use of the module, without doing the proper validations. And in this case I'd argue is better to reflect the desired responsability through the interface. The need to access the benchmark yaml is just an implementation detail of Snakemake, maybe with other workflow languages, we wouldn't need access to it.
2. Good point, asserts were leftover from my prototyping phase.
3. I added some more type hints and refactored some of the brittle code. It should be better now.
4. Unfortunately, I think most of the string manipulations are needed. They are an implementation detail of how snakemake magic works, and I tried as much as I can to contain them in a few private methods. Although I improved some of the parsing logic, hopefully its better now.

Commit:
Address PR comments

Commit:
Update omni-schema version to 0.0.2

Commit:
Module execution: clone repository, checkout commit, execute run.sh

Commit:
Add some logging

Commit:
Some small fixes with serialization of single node benchmark

Commit:
Fix workflow.run_node_workflow()

Commit:
Fix passing of arguments to Snakemake execution when running workflow

Commit:
Update module references for Benchmark_001.yaml

Commit:
Remove yaml converter

imallona commented on .gitignore:
```diff
@@ -316,4 +317,7 @@ playground/*
+.log/
+
+output_dag.png
+Snakefile
```
Could the Snakefile be part of the output folder? Not git tracked indeed, but easily reachable, e.g. if the out folder is `./out` then at `.out/Snakefile`

imallona replied:
ah, but it is already, solved, `def serialize_workflow(self, benchmark: Benchmark, output_dir: str = os.getcwd())` does this

DanInci replied:
Yes, Serialization of the snakefile can be output anywhere, by default is just the work directory.

imallona commented on src/workflow/snakemake/rules/rule_node.smk:
```diff
@@ -0,0 +1,84 @@
+from src.workflow.snakemake.format import formatter
+
+
+def create_node_rule(node, benchmark=None):
```
(future) perhaps this is the ideal place to extend with "extra" rule elements, e.g. envmodule, singularity, conda, log, benchmark etc

DanInci replied:
Added a comment in the file.

imallona commented on src/workflow/snakemake/rules/rule_start_benchmark.smk:
```diff
@@ -0,0 +1,19 @@
+import os
+
+rule start_benchmark:
```
(not urgent but) I think we could get rid of this. Or: better, replace by some archspec/nproc/OS fingerprinting, so we have some extra logging about the hardware used to run snakemake

DanInci replied:
I'll consider this as nice to have for now, would it be better to be included in a larger issue regarding logging?

imallona commented on src/workflow/snakemake/snakemake.py:
```diff
@@ -0,0 +1,190 @@
+
+        return success
+
+    def serialize_node_workflow(self, node: BenchmarkNode, output_dir: str = os.getcwd()):
```
[unstructured thoughts/rambling] I wonder how to align these viewpoints for ruleset rendering (benchmark vs node) as compared to the snakemake command calls themselves. So this is the most robust and user-friendly. I mean we generate the ruleset via omni_workflow, but then we execute it with a somewhat standardized version of `snakemake --cores x --use-singularity` or whatever. But we could also modify the snakemake call to `snakemake -R --until c --force --retries 200` (extreme case here). Perhaps combining snakemake command call arguments and ruleset generation patterns can provide extra features/fulfill some usecases.

imallona commented on src/workflow/snakemake/snakemake.py:
```diff
@@ -0,0 +1,190 @@
+        f.write("##\n")
+        f.write("## Snakefile to orchestrate YAML-defined omnibenchmarks\n")
+        f.write("##\n")
+        f.write(f"## This Snakefile has been automatically generated on {datetime.now()}\n")
```
(low prio) Worth incorporating the benchmarker name/email, and benchmark name and version? As parsed from the benchmarking yaml?

DanInci replied:
Great suggestion, I included your suggested information, should be easy to change / adapt in the future.
For now it will look like this:

```
#!/usr/bin/env snakemake -s
#############################################
# Snakefile for Orchestrating YAML-defined OmniBenchmarks
#############################################
# 
# This Snakefile has been automatically generated on 2024-07-22 16:42:56.973076
# 
# Benchmark Details:
# - Name: OmniBenchmark Benchmark_001
# - Version: 1.0
# 
# Author Information:
# - Contact: John Doe at Robinsons lab, john.doe@uzh.ch
#
#############################################
```

imallona reviewed (approved):
Looks good to me. I haven't tried to run the three PR/two branch octopus merge to see all changes talk nicely to each other, but did with `benchmark_model_updates` and worked beautifully

Commit:
Change interface for model.Benchmark such that it takes directly the benchmark_yaml path

DanInci committed:
Merge pull request #12 from omnibenchmark/remove_converter

Remove YAML Converter

Commit:
Some refactoring + include Almut's stage_order code

DanInci committed:
Merge pull request #11 from omnibenchmark/benchmark_model_updates

Benchmark model schema update + Repo cloning

Commit:
Include benchmark information in serialized Snakefile
